src/runtime-rs/crates/kata/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+        }
+        self.address = path
+            .to_str()
+            .map(|v| v.to_owned())
+            .ok_or(Error::InvalidArgument)?;
+
+        // Ensure `bundle` is a valid path.
+        if !self.bundle.is_empty() {
+            let path = PathBuf::from(self.bundle.clone())
+                .canonicalize()
+                .map_err(|_| Error::InvalidArgument)?;
+            let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+            if !md.is_dir() {
+                return Err(Error::InvalidArgument);
+            }
+            self.bundle = path
+                .to_str()
+                .map(|v| v.to_owned())
+                .ok_or(Error::InvalidArgument)?;
+        }
+
+        Ok(())
+    }
+
+    fn is_component_dangerous(comp: &str) -> bool {
+        if comp.is_empty() {
+            return true;
+        }
+        // only allow ascii alphanumeric character and '-', '_', '.' and '~'
+        !comp
+            .chars()
+            .all(|x| matches!(x, '0'..='9' | 'A'..='Z' | 'a'..='z' | '-' | '_' | '.' | '~'))
+    }
+}
+
+/// Command executor for shim.
+pub struct ShimExecutor {
+    args: ShimArgs,
+}
+
+impl ShimExecutor {
+    /// Create a new instance of [`Shim`].
+    pub fn new(args: ShimArgs) -> Self {
+        ShimExecutor { args }
+    }
+
+    fn get_bundle_path(&self) -> Result<PathBuf> {
+        std::env::current_dir().map_err(Error::BundlePath)
+    }
+
+    fn load_image_spec(&self) -> Result<oci_spec::runtime::Spec> {
+        let bundle_path = self.get_bundle_path()?;
+        let spec_file = bundle_path.join("config.json");
+
+        oci_spec::runtime::Spec::load(spec_file).map_err(Error::OciSpec)
+    }
+
+    fn write_address(&self, address: &Path) -> Result<()> {
+        let dir = self.get_bundle_path()?;
+        let file_path = &dir.join("address");
+        std::fs::write(file_path, address.as_os_str().as_bytes())
+            .map_err(|e| Error::WriteFile(e, file_path.clone()))
+    }
+
+    fn write_pid_file(&self, pid: &str) -> Result<()> {
+        let dir = self.get_bundle_path()?;
+        let file_path = &dir.join("shim.pid");
+        std::fs::write(file_path, pid.as_bytes())
+            .map_err(|e| Error::WriteFile(e, file_path.clone()))
+    }
+
+    fn socket_address(&self, id: &str) -> Result<PathBuf> {
+        if id.is_empty() {
+            return Err(Error::EmptySandboxId);
+        }
+        Ok(PathBuf::from(SOCKET_ROOT)
+            .join(&self.args.namespace)
+            .join(id)
+            .join(SHIM_SOCKET))
+    }
+
+    fn read_pid_file(&self, bundle_path: &Path) -> Result<String> {
+        let file_path = bundle_path.join("shim.pid");
+        // Limit the size of content to read in.
+        let size = file_path
+            .metadata()
+            .map_err(|e| Error::QueryFile(e, file_path.clone()))?
+            .len();
+        // 20 is the maximum number of decimal digits for u64
+        if size > 20 {
```

Comment:
I'd rather just try to convert the string value to a u64 and return an error if that fails. That would be clearer and safer imho. It would also assure you that what you read was actually a number and not some random junk (which currently isn't handled).